### PR TITLE
Log data type

### PR DESCRIPTION
**Pull request by saurabhd336**
Allows using CLP as a compression type for string columns. This hides all internal details of clp encoding, without having to expose the individual parts of a clp encoded log line externally. Giving a clean forwardIndex.getString() interface for the operators.

Enable using

```
"fieldConfigList": [
      {
        "name": "logLine",
        "encodingType": "RAW",
        "compressionCodec": "CLP",
      }

### CODE REVIEW

**Comment by saurabhd336** [comment]
cc: @kirkrodrigues @jackluo923 @npawar

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/CLPForwardIndexCreatorTest.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    logLines.add(
+        "2023/10/26 00:03:10.168 INFO [PropertyCache] [HelixController-pipeline-default-pinot-(4a02a32c_DEFAULT)] "
+            + "Event pinot::DEFAULT::4a02a32c_DEFAULT : Refreshed 35 property LiveInstance took 5 ms. Selective: true");
+    logLines.add(
+        "2023/10/26 00:03:10.169 INFO [PropertyCache] [HelixController-pipeline-default-pinot-(4a02a32d_DEFAULT)] "
+            + "Event pinot::DEFAULT::4a02a32d_DEFAULT : Refreshed 81 property LiveInstance took 4 ms. Selective: true");
+    logLines.add(
+        "2023/10/27 16:35:10.470 INFO [ControllerResponseFilter] [grizzly-http-server-2] Handled request from 0.0"
+            + ".0.0 GET https://0.0.0.0:8443/health?checkType=liveness, content-type null status code 200 OK");
+    logLines.add(
```
Comment (kirkrodrigues) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/CLPForwardIndexCreatorTest.java:70]: Can we include a test for a null row?

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java [hunk 2]
```diff
@@ -126,6 +126,9 @@ public enum CompressionCodec {
     SNAPPY(true, false),
     ZSTANDARD(true, false),
     LZ4(true, false),
+    // CLP is a special type of compression codec that isn't generally applicable to all RAW columns and has a
+    // special handling (see {@link CLPForwardIndexCreatorV1})
```
Comment (kirkrodrigues) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java:130]: > "a special handling"

It sounds like this is an incomplete sentence? Did you mean to say "special handling for log events" or something like that?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexCreatorFactory.java [hunk 3]
```diff
@@ -69,6 +71,10 @@ public static ForwardIndexCreator createIndexCreator(IndexCreationContext contex
     } else {
       // Dictionary disabled columns
       DataType storedType = fieldSpec.getDataType().getStoredType();
+      if (indexConfig.getCompressionCodec() == FieldConfig.CompressionCodec.CLP) {
+        // CLP compression codec
```
Comment (kirkrodrigues) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexCreatorFactory.java:75]: This comment seems a little redundant with the if condition.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/StringColumnPreIndexStatsCollector.java [hunk 4]
```diff
@@ -74,6 +91,14 @@ public void collect(Object entry) {
     }
   }
 
+  @Override
+  public CLPStats getCLPStats() {
+    if (_sealed) {
+      return _clpStatsCollector.getCLPStats();
+    }
+    throw new IllegalStateException("you must seal the collector first before asking for clp stats");
```
Comment (kirkrodrigues) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/StringColumnPreIndexStatsCollector.java:99]:
```suggestion
    throw new IllegalStateException("The collector must be sealed before calling getCLPStats");
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV1.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+  private void copyFileIntoBuffer(File file) throws IOException {
+    try (FileChannel from = (FileChannel.open(file.toPath(), StandardOpenOption.READ))) {
+      _fileBuffer.put(from.map(FileChannel.MapMode.READ_ONLY, 0, file.length()));
+    }
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    // Delete all temp file
```
Comment (kirkrodrigues) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV1.java:268]:
```suggestion
    // Delete all temp files
```

**Review by kirkrodrigues — COMMENTED** [review]
This is neat! I'm no Pinot expert, so I can't comment on the fine-grained validity of the code, but it roughly makes sense to me. I added a few minor suggestions. I would also suggest renaming any new classes as `ClpXXX` rather than `CLPXXX`, so that the classes play better with serialization/deserialization libraries (if ever necessary).

Perhaps this is a silly question, but how can we support CLP's query logic (like [clpMatch](https://github.com/apache/pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/ClpRewriter.java#L69)) on top of this?

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/CLPEncodingRealtimeIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.pinot.spi.config.table.ingestion.IngestionConfig;
+import org.apache.pinot.spi.config.table.ingestion.TransformConfig;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.util.TestUtils;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class CLPEncodingRealtimeIntegrationTest extends BaseClusterIntegrationTestSet {
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/CLPEncodingRealtimeIntegrationTest.java:37]: I feel an integration test is a little bit overkilling here since all we want to verify is reading from the forward index. Suggest removing this one and using unit test to cover it

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/ForwardIndexConfig.java [hunk 7]
```diff
@@ -58,6 +58,7 @@ public ForwardIndexConfig(@Nullable Boolean disabled, @Nullable CompressionCodec
     if (compressionCodec != null) {
       switch (compressionCodec) {
         case PASS_THROUGH:
+        case CLP:
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/ForwardIndexConfig.java:61]: This seems like a hack. Should we count CLP as a raw value compression? We don't use dictionary encoding on the values

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/CLPForwardIndexCreatorTest.java (hunk 1 above)
Comment (saurabhd336) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/CLPForwardIndexCreatorTest.java:70]: Added

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java (hunk 2 above)
Comment (saurabhd336) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java:130]: Modified

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexCreatorFactory.java (hunk 3 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexCreatorFactory.java:75]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/StringColumnPreIndexStatsCollector.java (hunk 4 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/StringColumnPreIndexStatsCollector.java:99]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV1.java (hunk 5 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV1.java:268]: Ack
